[nnc] Add LoopNest::unsafe_fuseLoops to let users apply fusion on stmts that may violate our correctness checks

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60601**

Differential Revision: [D29346128](https://our.internmc.facebook.com/intern/diff/D29346128)

huiguoo committed:
[nnc] Add LoopNest::unsafe_fuseLoops to let users apply fusion on stmts that may violate our correctness checks

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29346128).

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1615,6 +1615,37 @@ bool LoopNest::hasLoopCarriedDependence(For* loop) {
   return false;
 }
 
+bool LoopNest::unsafe_fuseLoops(const std::vector<For*>& loops, For** fused) {
```
I'm a bit skeptical about allowing fusion like this when it could be completely wrong to fuse. I understand this API is called `unsafe_` but still..

In addition to not checking for loop-carried dependence, this API does not check if the loops have the same parents, that the loops do not have any statements between them, that the loop bounds are the same. In all these cases, it is just wrong to fuse the loops.

What is the need for this API? Are you trying to have this just to cover the corner cases that the fusion API does not allow?

ZolotukhinM replied:
I think it's useful to have both "safe" and "unsafe" versions. In many cases the user might know that fusing two loops is legal where as it could be extremely hard to automatically prove that in compiler. When, say, we're constructing a loop nest for matmul or conv we pretty much might know what can be fused and what can't. The safe version is still important too and can be used in automatic transformations.

navahgar replied:
The only case we cannot prove that fusing is legal is when we report some loop-carried dependences which are false positives. I agree that, in those cases, we can allow "unsafe" fusion without checking for the loop-carried dependence.

We should definitely not allow fusing loops when other conditions are not satisfied, like loops having the same bounds, etc, even in the "unsafe" case.

ZolotukhinM replied:
Yeah, that makes total sense.

huiguoo replied:
In the "unsafe" version, I tried to remove as much checks as possible and leave the correctness issue to the user.

> We should definitely not allow fusing loops when other conditions are not satisfied, like loops having the same bounds, etc, even in the "unsafe" case.

 Consider symbolic expression bounds, our analysis is not capable to prove the equivalence of such bounds yet. For example, `2s/(s+1)` as loop1 bound and `1` as loop2 bound, `s` is a positive integer - our check does not deem the two equivalent yet.  We can extend our check to cover more cases but should unblock the user before that.

>  this API does not check if the loops have the same parents, that the loops do not have any statements between them

In my understanding, it is not necessary for the loops to have parents to fuse them, also, not necessary to not have any statements between them. Am I missing something?

huiguoo replied:
Ah OK, I think I understand the same parent condition. Will add a check for that. But still, the loops can have statements between them as long as there are no data dependence between these statements and the loops.

ZolotukhinM replied:
How about we disallow fusion if there stmts between loops (even unsafe version), but provide api to arbitrarily shuffle stmts within a block? It might be less convenient to use since it's broken down into more primitive steps, but it can lead to more composable api in the end.

huiguoo replied:
This sounds reasonable. Will make the changes!

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.h:
```diff
@@ -276,6 +276,10 @@ class TORCH_API LoopNest {
   //
+  // This transformation is unsafe as it simply add all loops into the body of
+  // the first loop for fusion without any correctness checks.
+  static bool unsafe_fuseLoops(const std::vector<For*>& loops, For** fused);
```
Nit: let's stay consistent with lowerCamelCaseNames.

huiguoo committed:
Update on "[nnc] Add LoopNest::unsafe_fuseLoops to let users apply fusion on stmts that may violate our correctness checks"


Differential Revision: [D29346128](https://our.internmc.facebook.com/intern/diff/D29346128)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29346128).

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1674,41 +1698,29 @@ bool LoopNest::fuseLoops(const std::vector<For*>& loops, For** fused) {
   // So, we create a clone of all the loops, fuse them and check for this.
   std::vector<For*> loops_copy;
   loops_copy.reserve(loops.size());
+  std::vector<Stmt*> loops_copy2;
```
You can avoid this vector by creating a `Block` before the loop below and add the statements to the block as the clones are created.

navahgar reviewed (commented):
Mostly LGTM. Just one minor comment.

huiguoo committed:
Update on "[nnc] Add LoopNest::unsafe_fuseLoops to let users apply fusion on stmts that may violate our correctness checks"


Differential Revision: [D29346128](https://our.internmc.facebook.com/intern/diff/D29346128)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29346128).

navahgar reviewed (approved):
LGTM
